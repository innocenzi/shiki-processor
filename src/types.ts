import type { HtmlRendererOptions } from 'shiki'

export type LineOptions = NonNullable<HtmlRendererOptions['lineOptions']>

export interface ProcessorResult {
	code: string
	lineOptions: LineOptions
}

export type PostProcessorResult = string | undefined

export interface TagClassesDictionary {
	[tag: string]: string[]
}

export interface ProcessorOptions {
	code: string
	lang: string
}

export type ProcessorHandler = (options: ProcessorOptions) => ProcessorResult
export type PostProcessorHandler = (options: ProcessorOptions) => PostProcessorResult

export interface Processor {
	name: string
	handler?: ProcessorHandler
	postProcess?: PostProcessorHandler
}
