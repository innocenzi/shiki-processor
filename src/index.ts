export * from './processors'
export * from './utils'
export * from './types'

export { process, postProcess } from './processor'
export { getHighlighter } from './highlighter'
