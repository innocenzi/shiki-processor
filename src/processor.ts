import type { Processor, ProcessorResult } from './types'

/**
 * Defines a processor.
 */
export function defineProcessor(processor: Processor): Processor {
	return processor
}

/**
 * Transforms code through the given processors.
 */
export function process(processors: Processor[], code: string, lang: string) {
	return processors.reduce((options, processor) => {
		const { code, lineOptions } = processor.handler?.({
			code: options.code,
			lang,
		}) ?? options

		return {
			code,
			lineOptions: [
				...options.lineOptions,
				...lineOptions,
			],
		}
	}, {
		code,
		lineOptions: [],
	} as ProcessorResult)
}

/**
 * Transforms final code through the given processors.
 */
export function postProcess(processors: Processor[], code: string, lang: string) {
	return processors.reduce((code, processor) => processor.postProcess?.({
		code,
		lang,
	}) ?? code, code)
}
