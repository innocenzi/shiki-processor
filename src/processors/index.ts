export { default as createDiffProcessor, DiffProcessorOptions } from './diff'
export { default as createFocusProcessor, FocusProcessorOptions } from './focus'
export { default as createHighlightProcessor, HighlightProcessorOptions } from './highlight'
