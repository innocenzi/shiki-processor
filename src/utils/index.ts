export * from './add-class'
export * from './create-range-processor'
