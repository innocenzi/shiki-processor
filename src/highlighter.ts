import { getHighlighter as getShikiHighlighter, Highlighter, HighlighterOptions } from 'shiki'
import { postProcess, process } from './processor'
import type { Processor } from './types'

interface Options extends HighlighterOptions {
	processors?: Processor[]
}

export async function getHighlighter(options: Options = {}): Promise<Highlighter> {
	const highlighter = await getShikiHighlighter(options)
	const processors = options.processors ?? []

	return {
		...highlighter,
		codeToHtml: (str, htmlOptions) => {
			const lang = typeof htmlOptions === 'object'
				? htmlOptions.lang!
				: htmlOptions!

			const baseLineOptions = typeof htmlOptions === 'object'
				? htmlOptions.lineOptions ?? []
				: []

			const theme = typeof htmlOptions === 'object'
				? htmlOptions.theme
				: undefined

			const { code, lineOptions } = process(processors, str, lang)

			const highlighted = highlighter.codeToHtml(code, {
				lang,
				theme,
				lineOptions: [
					...lineOptions,
					...baseLineOptions,
				],
			})

			return postProcess(processors, highlighted, lang)
		},
	}
}
