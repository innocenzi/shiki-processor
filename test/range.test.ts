import { expect, it } from 'vitest'
import { createRangeProcessor, getHighlighter } from '../src'

it('understands the range syntax', async() => {
	const highlighter = await getHighlighter({
		processors: [{
			name: 'test',
			handler: createRangeProcessor({
				test: ['is-tested'],
			}),
		}],
	})

	const snippet = `
	function () {
		console.log('uwu')  // [!code test:2]
		console.log('owo')
	}
	`

	expect(highlighter.codeToHtml(snippet, { lang: 'javascript' })).toMatchSnapshot()
})
