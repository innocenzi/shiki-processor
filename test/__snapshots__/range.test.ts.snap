// Vitest Snapshot v1

exports[`understands the range syntax 1`] = `
"<pre class=\\"shiki\\" style=\\"background-color: #2e3440ff\\"><code><span class=\\"line\\"></span>
<span class=\\"line\\"><span style=\\"color: #D8DEE9FF\\">	</span><span style=\\"color: #81A1C1\\">function</span><span style=\\"color: #D8DEE9FF\\"> </span><span style=\\"color: #ECEFF4\\">()</span><span style=\\"color: #D8DEE9FF\\"> </span><span style=\\"color: #ECEFF4\\">{</span></span>
<span class=\\"line is-tested\\"><span style=\\"color: #D8DEE9FF\\">		</span><span style=\\"color: #D8DEE9\\">console</span><span style=\\"color: #ECEFF4\\">.</span><span style=\\"color: #88C0D0\\">log</span><span style=\\"color: #D8DEE9FF\\">(</span><span style=\\"color: #ECEFF4\\">&#39;</span><span style=\\"color: #A3BE8C\\">uwu</span><span style=\\"color: #ECEFF4\\">&#39;</span><span style=\\"color: #D8DEE9FF\\">)  </span></span>
<span class=\\"line is-tested\\"><span style=\\"color: #D8DEE9FF\\">		</span><span style=\\"color: #D8DEE9\\">console</span><span style=\\"color: #ECEFF4\\">.</span><span style=\\"color: #88C0D0\\">log</span><span style=\\"color: #D8DEE9FF\\">(</span><span style=\\"color: #ECEFF4\\">&#39;</span><span style=\\"color: #A3BE8C\\">owo</span><span style=\\"color: #ECEFF4\\">&#39;</span><span style=\\"color: #D8DEE9FF\\">)</span></span>
<span class=\\"line\\"><span style=\\"color: #D8DEE9FF\\">	</span><span style=\\"color: #ECEFF4\\">}</span></span>
<span class=\\"line\\"><span style=\\"color: #D8DEE9FF\\">	</span></span></code></pre>"
`;
